### PR TITLE
refactor(rpc/v0.7): serializer with version awareness

### LINKED ISSUE #1830: RPC: version-aware serialization
Our current RPC upgrades are painful as specification diffs cannot be trivially traced to impacted methods and DTOs. Create a new version aware serializer which enables this.

Implement this for RPC v0.7 types and methods. We can then work backwards from this basis as required.

-------------------
Components from v0.7 starknet api. Some of these are input only and don't need to be implemented.

- [x] EVENTS_CHUNK
- [ ] RESULT_PAGE_REQUEST
- [x] EMITTED_EVENT
- [x] EVENT
- [x] EVENT_CONTENT
- [ ] EVENT_FILTER
- [ ] BLOCK_ID
- [ ] BLOCK_TAG
- [x] SYNC_STATUS
- [x] NUM_AS_HEX
- [ ] u64
- [ ] u128
- [x] CHAIN_ID
- [x] STATE_DIFF
- [x] PENDING_STATE_UPDATE
- [x] STATE_UPDATE
- [x] ADDRESS
- [ ] STORAGE_KEY
- [x] ETH_ADDRESS
- [x] TXN_HASH
- [x] FELT
- [x] BLOCK_NUMBER
- [x] BLOCK_HASH
- [ ] BLOCK_BODY_WITH_TX_HASHES
- [ ] BLOCK_BODY_WITH_TXS
- [ ] BLOCK_BODY_WITH_RECEIPTS
- [ ] BLOCK_HEADER
- [ ] PENDING_BLOCK_HEADER
- [ ] BLOCK_WITH_TX_HASHES
- [ ] BLOCK_WITH_TXS
- [ ] BLOCK_WITH_RECEIPTS
- [ ] PENDING_BLOCK_WITH_TX_HASHES
- [ ] PENDING_BLOCK_WITH_TXS
- [ ] PENDING_BLOCK_WITH_RECEIPTS
- [x] DEPLOYED_CONTRACT_ITEM
- [x] CONTRACT_STORAGE_DIFF_ITEM
- [ ] TXN
- [ ] SIGNATURE
- [ ] DECLARE_TXN
- [ ] DECLARE_TXN_V0
- [ ] DECLARE_TXN_V1
- [ ] DECLARE_TXN_V2
- [ ] DECLARE_TXN_V3
- [ ] BROADCASTED_TXN
- [ ] BROADCASTED_INVOKE_TXN
- [ ] BROADCASTED_DEPLOY_ACCOUNT_TXN
- [ ] BROADCASTED_DECLARE_TXN
- [ ] BROADCASTED_DECLARE_TXN_V1
- [ ] BROADCASTED_DECLARE_TXN_V2
- [ ] BROADCASTED_DECLARE_TXN_V3
- [ ] DEPLOY_ACCOUNT_TXN
- [ ] DEPLOY_ACCOUNT_TXN_V1
- [ ] DEPLOY_ACCOUNT_TXN_V3
- [ ] DEPLOY_TXN
- [ ] INVOKE_TXN_V0
- [ ] INVOKE_TXN_V1
- [ ] INVOKE_TXN_V3
- [ ] INVOKE_TXN
- [ ] L1_HANDLER_TXN
- [x] COMMON_RECEIPT_PROPERTIES
- [x] INVOKE_TXN_RECEIPT
- [ ] PENDING_INVOKE_TXN_RECEIPT
- [x] DECLARE_TXN_RECEIPT
- [ ] PENDING_DECLARE_TXN_RECEIPT
- [x] DEPLOY_ACCOUNT_TXN_RECEIPT
- [ ] PENDING_DEPLOY_ACCOUNT_TXN_RECEIPT
- [x] DEPLOY_TXN_RECEIPT
- [x] L1_HANDLER_TXN_RECEIPT
- [ ] PENDING_L1_HANDLER_TXN_RECEIPT
- [x] TXN_RECEIPT
- [ ] PENDING_TXN_RECEIPT
- [ ] PENDING_COMMON_RECEIPT_PROPERTIES
- [x] MSG_TO_L1
- [ ] MSG_FROM_L1
- [ ] TXN_STATUS
- [ ] TXN_FINALITY_STATUS
- [ ] TXN_EXECUTION_STATUS
- [ ] TXN_TYPE
- [ ] BLOCK_STATUS
- [ ] FUNCTION_CALL
- [x] CONTRACT_CLASS
- [x] DEPRECATED_CONTRACT_CLASS
- [x] DEPRECATED_CAIRO_ENTRY_POINT
- [x] SIERRA_ENTRY_POINT
- [x] CONTRACT_ABI
- [x] CONTRACT_ABI_ENTRY
- [x] STRUCT_ABI_TYPE
- [x] EVENT_ABI_TYPE
- [x] FUNCTION_ABI_TYPE
- [x] STRUCT_ABI_ENTRY
- [x] STRUCT_MEMBER
- [x] EVENT_ABI_ENTRY
- [x] FUNCTION_STATE_MUTABILITY
- [x] FUNCTION_ABI_ENTRY
- [x] TYPED_PARAMETER
- [ ] SIMULATION_FLAG_FOR_ESTIMATE_FEE
- [x] PRICE_UNIT
- [ ] FEE_ESTIMATE
- [ ] FEE_PAYMENT
- [ ] DA_MODE
- [ ] RESOURCE_BOUNDS_MAPPING
- [ ] RESOURCE_BOUNDS
- [ ] RESOURCE_PRICE
- [x] COMPUTATION_RESOURCES
- [x] EXECUTION_RESOURCES
- [ ] FAILED_TO_RECEIVE_TXN

-----------------------

Components from v0.7 trace API

- [ ] TRANSACTION_TRACE
- [ ] SIMULATION_FLAG
- [ ] NESTED_CALL
- [ ] FUNCTION_INVOCATION
- [ ] ENTRY_POINT_TYPE
- [ ] CALL_TYPE
- [ ] ORDERED_EVENT
- [ ] ORDERED_MESSAGE

-----------------------

Methods from v0.7

- [x] starknet_blockHashAndNumber
- [x] starknet_blockNumber
- [x] starknet_chainId
- [x] starknet_getBlockTransactionCount
- [x] starknet_getClass
- [x] starknet_getClassAt
- [x] starknet_getClassHashAt
- [x] starknet_getNonce
- [x] starknet_getStorageAt
- [x] starknet_getEvents
- [x] starknet_getStateUpdate
- [x] starknet_syncing
- [ ] starknet_call
- [ ] starknet_getTransactionStatus
- [ ] starknet_addDeclareTransaction
- [ ] starknet_addDeployAccountTransaction
- [ ] starknet_addInvokeTransaction
- [ ] starknet_getTransactionByBlockIdAndIndex
- [ ] starknet_getTransactionByHash
- [ ] starknet_estimateFee
- [ ] starknet_estimateMessageFee
- [ ] starknet_getBlockWithTxHashes
- [ ] starknet_getBlockWithTxs
- [x] starknet_getTransactionReceipt
- [ ] starknet_simulateTransactions
- [x] starknet_specVersion
- [ ] starknet_traceBlockTransactions
- [ ] starknet_traceTransaction
- [ ] starknet_getBlockWithReceipts

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
This PR adds introduces a custom serializer for RPC DTOs which can take the RPC version into account.

This will allow our code diffs to more closely match the actual specification diff; hopefully leading to much simpler version updates and reviews. And less bugs.

This PR implements this new serialization for all DTOs as per v0.7 of the specification.

I'm opening this early so reviews can start even though I anticipate this will still take a few days to complete. I promise I won't rebase -- reviews can therefore be paused/continued at any commit.

### How to review

This is an extremely large PR. Thankfully it is completely orthogonal to all other work, so we can take our time.

I recommend reviewing it by commit. Almost all commits contain a single DTO implementation which means you can review it by opening the specification at that DTO and compare them.

### What to check

Both the implementation and test for a DTO should match the types used in the specification. e.g. if a DTO `T` has `T.y: Felt` in the spec, then the implementation should use `serialize(&Felt(t.y))` and not implement `Felt` manaully.

In other words, serialization should use DTOs as per the spec, and not inline them (unless of course they're inlined in the spec definition).

Similarly, tests should re-use the DTOs as well. Don't inline the DTOs since that would cause breakage when the spec changes.

### Aftermath

Once I've completed; we should still

- [ ] benchmark performance for current methods (since they now use the new trait)
- [ ] benchmark new methods
- [ ] consider squashing commits into groups e.g. state update DTOs, block DTOs

Closes #1830 -- you can also check this for a list of DTOs and progress.

### CODE REVIEW

**Comment by Mirko-von-Leipzig** [comment]
> If the RPC tests still pass, we can be positive that nothing's broken, right?
> 
> LGTM.

Not entirely; methdos aren't actively updated to using these DTOs yet. And likely won't be for quite a while since you need almost all of the DTOs implemented for every version which will take ages. We can say more once we start switching methods over, but we'll want to be selective initially.

All we can be certain of is that switching the serializer implementation under-the-hood didn't break anything.

**Comment by Mirko-von-Leipzig** [comment]
Closing in favor of splitting into multiple smaller PRs.
